Converge all use of EHcache to v3

mwoodiupui opened the pull request:
## References
* Fixes #8099

## Description
Replace EHcache v2 with v3 wherever we use EHcache.

## Instructions for Reviewers

List of changes in this PR:
* Remove the unused CachingService
* Reconfigure Hibernate to use its JSR-107 (JCache) cache provider instead of EHcache.
* Rewrite the Hibernate EHcache configuration for v3
* Disable the disk caching tier because it causes mysterious, fatal exceptions during startup.

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

## Checklist
- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

mwoodiupui committed:
Switch Hibernate to JCache #8099

mwoodiupui committed:
Begin moving CachingService to Ehcache v3 # 8099

mwoodiupui committed:
Rip out the unused CachingService.  $8099

mwoodiupui committed:
Rewrite Hibernate Ehcache configuration for v3.  #8099

Also fix incorrectly-used Spring-style <property> elements in hibernate.cfg.xml

mwoodiupui committed:
Merge remote-tracking branch 'upstream/main' into 8099

mwoodiupui commented:
I've tried to re-create the Hibernate cache configuration as best I could, but v2 and v3 are quite different and some features were not carried over.

Adding a disk tier causes initialization errors which seem to be related to direct-mapped buffers.  The stack trace was absolutely no help.  I've disabled the disk tier (but left my best guess at its configuration commented out).  I'm not sure that overflow to disk makes much sense, given that we are caching things we can get from the database.  I've left the on-heap caches fairly generous and hope that that will be sufficient.  Discussion of cache sizing, and especially where the original numbers come from, would be welcome.

tdonohue reviewed (changes requested):
@mwoodiupui : Gave this a test today, and unfortunately I cannot get the `server` webapp to boot up with these changes in place.  I'm getting this error:
```
14-Feb-2022 21:01:05.077 SEVERE [main] org.apache.catalina.startup.HostConfig.deployDirectory Error deploying web application directory [/usr/local/tomcat/webapps/server]
java.lang.IllegalStateException: Error starting child
....
Caused by: org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[/server]]
...
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'authenticationRestController': Unsatisfied dependency expressed through field 'converter'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'converterService': Unsatisfied dependency expressed through field 'webSecurityExpressionEvaluator'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'webSecurityExpressionEvaluator' defined in file [/dspace/webapps/server/WEB-INF/classes/org/dspace/app/rest/security/WebSecurityExpressionEvaluator.class]: Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'springSecurityFilterChain' defined in class path resource [org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.servlet.Filter]: Factory method 'springSecurityFilterChain' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'mvcHandlerMappingIntrospector' defined in class path resource [org/springframework/boot/autoconfigure/web/servlet/WebMvcAutoConfiguration$EnableWebMvcConfiguration.class]: Invocation of init method failed; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'restHandlerMapping' defined in class path resource [org/springframework/data/rest/webmvc/config/RepositoryRestMvcConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.data.rest.webmvc.support.DelegatingHandlerMapping]: Factory method 'restHandlerMapping' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jpaHelper' defined in class path resource [org/springframework/data/rest/webmvc/config/RepositoryRestMvcConfiguration.class]: Initialization of bean failed; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'sessionFactory' defined in file [/dspace/config/spring/api/core-hibernate.xml]: Invocation of init method failed; nested exception is java.lang.NoClassDefFoundError: net/bytebuddy/NamingStrategy$SuffixingRandom$BaseNameResolver
...
 Caused by: java.lang.NoClassDefFoundError: net/bytebuddy/NamingStrategy$SuffixingRandom$BaseNameResolver
           at org.hibernate.cfg.Environment.buildBytecodeProvider(Environment.java:345)
           at org.hibernate.cfg.Environment.buildBytecodeProvider(Environment.java:337)
           at org.hibernate.cfg.Environment.<clinit>(Environment.java:230)
           at org.hibernate.boot.registry.StandardServiceRegistryBuilder.<init>(StandardServiceRegistryBuilder.java:122)
           at org.hibernate.boot.registry.StandardServiceRegistryBuilder.<init>(StandardServiceRegistryBuilder.java:96)
           at org.hibernate.cfg.Configuration.reset(Configuration.java:158)
           at org.hibernate.cfg.Configuration.<init>(Configuration.java:130)
```

So, it appears Hibernate is failing cause it cannot locate (at least one) ByteBuddy class.

My best guess is it's the exclusion noted below that's to blame here.

I gave this a quick code review & overall it looks good to me.  The default settings you applied also seem reasonable, especially since they look like they are directly mapped over from the older configuration for EHCache v2.

So, once the startup issue is fixed, I'll gladly re-test and see how things look.  Thanks!

mwoodiupui committed:
Remove outdated exclusion.  #8099

Also add some missing direct dependencies.  I thought these were
causing a problem, which in fact came from an old version of httpcore,
but we need these anyway.

mwoodiupui commented on dspace-api/pom.xml:
```diff
@@ -336,16 +338,30 @@
+                <exclusion>
+                    <groupId>net.bytebuddy</groupId>
+                    <artifactId>byte-buddy</artifactId>
+                </exclusion>
```
Thanks for tracking this down!  I just removed the exclusion.  It seems that Mockito no longer pulls in `byte-buddy`.

This is why nowadays I'll usually pin down a version in `<dependencyManagement>` rather than use an exclusion.

tdonohue reviewed (changes requested):
@mwoodiupui : Tried this again today.  The `/server` webapp now boots up, but I'm getting caching errors in the Tomcat log files when Spring Boot starts.  These errors don't seem to affect the `/server` webapp, but seem to be related to the old `/rest` webapp.  

Here's what I see just after Spring Boot starts in the Tomcat logs:

```
17:38:29.999 [main] ERROR org.ehcache.core.EhcacheManager - Initialize failed.
17:38:30.101 [main] ERROR org.ehcache.core.EhcacheManager - Initialize failed.
17:38:30.102 [main] ERROR org.dspace.servicemanager.DSpaceServiceManager - Error creating bean with name 'org.dspace.core.DBConnection': Unsatisfied dependency expressed through field 'sessionFactory'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'sessionFactory' defined in file [/dspace/config/spring/api/core-hibernate.xml]: Invocation of init method failed; nested exception is org.hibernate.service.spi.ServiceException: Unable to create requested service [org.hibernate.cache.spi.CacheImplementor]
...
 Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'sessionFactory' defined in file [/dspace/config/spring/api/core-hibernate.xml]: Invocation of init method failed; nested exception is org.hibernate.service.spi.ServiceException: Unable to create requested service [org.hibernate.cache.spi.CacheImplementor]
...
 Caused by: org.hibernate.service.spi.ServiceException: Unable to create requested service [org.hibernate.cache.spi.CacheImplementor]
...
 Caused by: java.lang.IllegalStateException: Cache provider not started
...
 Caused by: org.ehcache.StateTransitionException: Persistence directory already locked by this process: /usr/local/tomcat/temp
...
 Caused by: java.lang.RuntimeException: Persistence directory already locked by this process: /usr/local/tomcat/temp
...
 Caused by: java.nio.channels.OverlappingFileLockException
   at sun.nio.ch.FileLockTable.checkList(FileLockTable.java:229) ~[?:?]
   at sun.nio.ch.FileLockTable.add(FileLockTable.java:123) ~[?:?]
   at sun.nio.ch.FileChannelImpl.tryLock(FileChannelImpl.java:1154) ~[?:?]
   at java.nio.channels.FileChannel.tryLock(FileChannel.java:1165) ~[?:?]
   at org.ehcache.impl.persistence.DefaultLocalPersistenceService.internalStart(DefaultLocalPersistenceService.java:93) ~[ehcache-3.4.0.jar:3.4.0 45a1f99e0d65f13bdd379e60a5ca9a7b6d3a90bc]
...
17:38:30.110 [main] FATAL org.dspace.core.Context - Cannot obtain the bean which provides a database connection. Check previous entries in the dspace.log to find why the db failed to initialize.
17-Feb-2022 17:38:30.118 SEVERE [main] org.apache.catalina.core.StandardContext.startInternal One or more listeners failed to start. Full details will be found in the appropriate container log file
17-Feb-2022 17:38:30.119 SEVERE [main] org.apache.catalina.core.StandardContext.startInternal Context [/rest] startup failed due to previous errors
```

It **appears** to me that this error may be coming from the old, obsolete `/rest` webapp (which is deployed by default in my local Docker setup).  So, it may be that webapp has not been fully updated for EHCache v3.  

In any case, even with these errors, the `/server` webapp appears to function properly, and I'm not setting errors in the `dspace.log` when I use the `/server` webapp via the Angular UI.

So, I think this PR is getting closer...but we need to figure out a way to fix the old `/rest` webapp if possible.

mwoodiupui committed:
Remove all traces of disk cache to eliminate cache directory collisions. #8099

mwoodiupui commented:
I think the problem is that lingering persistence settings caused multiple concurrent webapps to compete for the same persistence directory.  Since we have no persistent caches, the simplest fix is to remove the <persistence/> element.

tdonohue reviewed (approved):
👍 Thanks @mwoodiupui !  This all looks great now and works well for me.  Everything now spins up & works (including the deprecated old REST API). I also see the new caches as INFO statements in the `dspace.log` when I start DSpace.

So, I think this is ready to merge.  I'll merge this ASAP, as we have several dependency update PRs open at once...so, best to start merging these & getting any conflicts worked out immediately.

Thanks again for your hard work on getting this EHCache stuff cleaned up.
